qiita_client/plugin.py

```diff
@@ -0,0 +1,330 @@
+            qclient.complete_job(job_id, success, error_msg=error_msg,
+                                 artifacts_info=artifacts_info)
+
+
+class QiitaTypePlugin(BaseQiitaPlugin):
+    """Represents a Qiita Type Plugin
+
+    Parameters
+    ----------
+    name : str
+        The plugin name
+    version : str
+        The plugin version
+    description : str
+        The plugin description
+    validate_func : callable
+        The function used to validate artifacts
+    html_generator_func : callable
+        The function used to generate the HTML generator
+    artifact_types : list of QiitaArtifactType
+        The artifact types defined in this plugin
+
+    Notes
+    -----
+    Both `validate_func` and `html_generator_func` should be a callable
+    that conforms to the signature:
+    function(qclient, job_id, job_parameters, output_dir)
+    where qclient is an instance of QiitaClient, job_id is a string with
+    the job identifier, job_parameters is a dictionary with the parameters
+    of the command and output_dir is a string with the output directory
+    """
+    _plugin_type = "artifact definition"
+
+    def __init__(self, name, version, description, validate_func,
+                 html_generator_func, artifact_types, publications=None):
+        super(QiitaTypePlugin, self).__init__(name, version, description,
+                                              publications=publications)
+
+        self.artifact_types = artifact_types
+
+        val_cmd = QiitaCommand(
+            'Validate', 'Validates a new artifact', validate_func,
+            {'template': ('prep_template', None),
+             'files': ('string', None),
+             'artifact_type': ('string', None)}, {}, None)
+
+        self._register_command(val_cmd)
+
+        html_cmd = QiitaCommand(
+            'Generate HTML summary', 'Generates the HTML summary',
+            html_generator_func,
+            {'input_data': ('artifact',
+                            [a.name for a in self.artifact_types])}, {}, None)
+
+        self._register_command(html_cmd)
+
+    def _install(self, qclient):
+        """Installs the plugin in Qiita"""
+        for at in self.artifact_types:
+            data = {'type_name': at.name,
+                    'description': at.description,
+                    'can_be_submitted_to_ebi': at.ebi,
+                    'can_be_submitted_to_vamps': at.vamps,
+                    'filepath_types': dumps(at.fp_types)}
+            qclient.post('/qiita_db/artifacts/types/', data=data)
+
+        super(QiitaTypePlugin, self)._install(qclient)
+
+
+class QiitaPlugin(BaseQiitaPlugin):
+    """Represents a Qiita Plugin"""
+
+    _plugin_type = "artifact transformation"
+
+    def register_command(self, command):
+        """Registers a command in the plugin
+
+        Parameters
+        ----------
+        command: QiitaCommand
+            The command to be added to the plugin
+        """
+        self._register_command(command)
```

Comment:
I don't want the `_register_command` function to be public on the `QiitaTypePlugin` class (although I'm using it internally). The rationale is that the commands in the `QiitaTypePlugin` are controlled (there are only 2 - 'Validate' and 'Generate HTML summary'), while `QiitaPlugin` can have any number of commands. However, both are actually represented in the same way, and that's why most of the functionality is added in the base class.